lnrpc/walletrpc/walletkit_server.go

```diff
@@ -819,3 +840,226 @@ func (w *WalletKit) LabelTransaction(ctx context.Context,
+// by passing in a raw TxTemplate message. If there are no inputs specified in
+// the template, coin selection is performed automatically. If the template does
+// contain any inputs, it is assumed that full coin selection happened
+// externally and no additional inputs are added. If the specified inputs aren't
+// enough to fund the outputs with the given fee rate, an error is returned.
+// After either selecting or verifying the inputs, all input UTXOs are locked
+// with an internal app ID.
+func (w *WalletKit) FundPsbt(_ context.Context,
+	req *FundPsbtRequest) (*FundPsbtResponse, error) {
+
+	var (
+		err         error
+		packet      *psbt.Packet
+		feeSatPerKW chainfee.SatPerKWeight
+	)
+
+	// There are two ways a user can specify what we call the template (a
+	// list of inputs and outputs to use in the PSBT): Either as a PSBT
+	// packet directly or as a special RPC message. Find out which one the
+	// user wants to use, they are mutually exclusive.
+	switch {
+	// The template is specified as a PSBT. All we have to do is parse it.
+	case req.GetPsbt() != nil:
+		r := bytes.NewReader(req.GetPsbt())
+		packet, err = psbt.NewFromRawBytes(r, false)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse PSBT: %v", err)
+		}
+
+	// The template is specified as a RPC message. We need to create a new
+	// PSBT and copy the RPC information over.
+	case req.GetRaw() != nil:
+		tpl := req.GetRaw()
+		if len(tpl.Outputs) == 0 {
+			return nil, fmt.Errorf("no outputs specified")
+		}
+
+		txOut := make([]*wire.TxOut, len(tpl.Outputs))
+		for idx, out := range tpl.Outputs {
+			txOut[idx] = &wire.TxOut{
+				Value:    out.Value,
+				PkScript: out.PkScript,
+			}
+		}
+
+		txIn := make([]*wire.OutPoint, len(tpl.Utxos))
+		for idx, in := range tpl.Utxos {
+			op, err := unmarshallOutPoint(in)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing "+
+					"outpoint: %v", err)
+			}
+			txIn[idx] = op
+		}
+
+		packet, err = psbt.New(txIn, txOut, 2, 0, nil)
+		if err != nil {
+			return nil, fmt.Errorf("could not create PSBT: %v", err)
+		}
+
+	default:
+		return nil, fmt.Errorf("transaction template missing, need " +
+			"to specify either PSBT or raw TX template")
+	}
+
+	// In case the user did specify inputs, we need to make sure they are
+	// known to us, still unspent and not yet locked.
+	if len(packet.UnsignedTx.TxIn) > 0 {
+		// Get a list of all unspent witness outputs. We'll acquire the
+		// global coin selection lock to ensure there aren't any other
+		// concurrent processes attempting to lock any UTXOs which may
+		// be shown available to us.
+		var utxos []*lnwallet.Utxo
+		err = w.cfg.CoinSelectionLocker.WithCoinSelectLock(func() error {
+			utxos, err = w.cfg.Wallet.ListUnspentWitness(
+				defaultMinConf, defaultMaxConf,
+			)
+			return err
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// Validate all inputs against our known list of UTXOs now.
```

Comment:
If inputs in the _PSBT_ were specified, then wouldn't we expect the input info to be populated as well? In this case, we can just do a sanity check that all the inputs are ours (though I think we should also be able to support only a subset of the inputs actually being ours).

Reply:
Optimally, yes. But because the wallet currently only supports adding own inputs, it populates/overwrites the input info anyway. We might change this in the future when adding foreign inputs is supported.